certdepot/depot_test.go

```diff
@@ -0,0 +1,473 @@
+						assert.Equal(t, string(certReqData), u.CertReq)
+						assert.Equal(t, string(certRevocListData), u.CertRevocList)
+						assert.Equal(t, newTTL, u.TTL)
+					},
+				},
+				{
+					name: "CheckOnExistingUserWithNoData",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "alice"
+						u := &User{
+							ID: name,
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						assert.False(t, d.Check(depot.CrtTag(name)))
+						assert.False(t, d.Check(depot.PrivKeyTag(name)))
+						assert.False(t, d.Check(depot.CsrTag(name)))
+						assert.False(t, d.Check(depot.CrlTag(name)))
+					},
+				},
+				{
+					name: "GetOnExistingUserWithNoData",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+						u := &User{
+							ID:  name,
+							TTL: time.Now(),
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						data, err := d.Get(depot.CrtTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.PrivKeyTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.CsrTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.CrlTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+					},
+				},
+				{
+					name: "GetOnExpiredTTL",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+						u := &User{
+							ID:            name,
+							Cert:          "cert",
+							PrivateKey:    "key",
+							CertReq:       "certReq",
+							CertRevocList: "certRevocList",
+							TTL:           time.Time{},
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						data, err := d.Get(depot.CrtTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.PrivKeyTag(name))
+						assert.NoError(t, err)
+						assert.Equal(t, u.PrivateKey, string(data))
+						data, err = d.Get(depot.CsrTag(name))
+						assert.NoError(t, err)
+						assert.Equal(t, u.CertReq, string(data))
+						data, err = d.Get(depot.CrlTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+					},
+				},
+
+				{
+					name: "DeleteWhenDNE",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+
+						assert.NoError(t, d.Delete(depot.CrtTag(name)))
+						assert.NoError(t, d.Delete(depot.PrivKeyTag(name)))
+						assert.NoError(t, d.Delete(depot.CsrTag(name)))
+						assert.NoError(t, d.Delete(depot.CrlTag(name)))
+					},
+				},
+			},
+		},
+	} {
+		t.Run(impl.name, func(t *testing.T) {
+			for _, test := range impl.tests {
+				t.Run(test.name, func(t *testing.T) {
+					d := impl.setup()
+					defer impl.cleanup()
+
+					test.test(t, d)
+				})
+			}
+			t.Run("Put", func(t *testing.T) {
+				d := impl.setup()
+				defer impl.cleanup()
+				name := "bob"
+
+				assert.Error(t, d.Put(depot.CrtTag(name), nil))
```

Comment:
should this be a require rather than an assert? 

require means that execution would stop here, and assert means that execution will continue even if the assertion fails. Often we overuse asserts and end up with way too much error output when a test fails. The risk of require is that one failure will mask another orthogonal failure... So it's fine to use asserts (and we certianly do a lot in different places, but...

Reply:
Yeah I try to put requires only when I know the following code will either panic or result in a ton of failing asserts. Sometimes it is questionable, but I think here nothing will panic and it would be helpful to see if the following code works despite this code not returning an error when it should.